src/api.js

```diff
@@ -0,0 +1,797 @@
+                        curresult = {
+                            "columns": stmt["getColumnNames"](),
+                            "values": []
+                        };
+                        results.push(curresult);
+                    }
+                    curresult["values"].push(stmt["get"]());
+                }
+            } finally {
+                stmt["free"]();
+            }
+        }
+        return results;
+    } finally {
+        stackRestore(stack);
+    }
+};
+
+/* Execute an sql statement, and call a callback for each row of result.
+
+**Currently** this method is synchronous, it will not return until the callback has
+been called on every row of the result. But this might change.
+
+@param sql [String] A string of SQL text. Can contain placeholders that will be
+bound to the parameters given as the second argument
+@param params [Array<String,Number,null,Uint8Array>] (*optional*) Parameters to bind
+to the query
+@param callback [Function(Object)] A function that will be called on each row of result
+@param done [Function] A function that will be called when all rows have been retrieved
+
+@return [Database] The database object. Useful for method chaining
+
+@example Read values from a table
+    db.each("SELECT name,age FROM users WHERE age >= $majority",
+                    {$majority:18},
+                    function(row){console.log(row.name + " is a grown-up.")}
+                );
+ */
+Database.prototype["each"] = function(sql, params, callback, done) {
+    var stmt;
+    if (typeof params === "function") {
+        done = callback;
+        callback = params;
+        params = void 0;
+    }
+    stmt = this["prepare"](sql, params);
+    try {
+        while (stmt["step"]()) {
+            callback(stmt["getAsObject"]());
+        }
+    } finally {
+        stmt["free"]();
+    }
+    if (typeof done === "function") {
+        return done();
+    }
+};
+
+/* Prepare an SQL statement
+@param sql [String] a string of SQL, that can contain placeholders ('?', ':VVV', ':AAA', '@AAA')
+@param params [Array] (*optional*) values to bind to placeholders
+@return [Statement] the resulting statement
+@throw [String] SQLite error
+ */
+Database.prototype["prepare"] = function(sql, params) {
+    var pStmt, stmt;
+    setValue(apiTemp, 0, "i32");
+    this.handleError(sqlite3_prepare_v2(this.db, sql, -1, apiTemp, NULL));
+    pStmt = getValue(apiTemp, "i32");
+    if (pStmt === NULL) {
+        throw "Nothing to prepare";
+    }
+    stmt = new Statement(pStmt, this);
+    if (params != null) {
+        stmt.bind(params);
+    }
+    this.statements[pStmt] = stmt;
+    return stmt;
+};
+
+/* Exports the contents of the database to a binary array
+@return [Uint8Array] An array of bytes of the SQLite3 database file
+ */
+Database.prototype["export"] = function() {
+    var _, binaryDb, func, ref, ref1, stmt;
+    ref = this.statements;
+    for (_ in ref) {
+        stmt = ref[_];
+        stmt["free"]();
+    }
```

Comment:
```suggestion
    for (var stmt in this.statements) this.statements[stmt]["free"]();
```

Reply:
for-in loops are forbidden in jslint (and are slow/expensive operations for historical reasons).  to pass jslint:

```diff
-ref = this.statements;
-for (_ in ref) {
-    stmt = ref[_];
-    stmt["free"]();
-}
+ // Object.values() is better, but lets stick with es5 for maximum compatibility
+ Object.keys(this.statements).forEach(function (key) {
+    this.statements[key].free();
+});

...

+ // bug-workaround for closure renaming properties
+ /* jslint ignore:start */
+ Database.prototype["close"] = Database.prototype.close;
+ Database.prototype["export"] = Database.prototype.export;
+ Database.prototype["free"] = Database.prototype.free;
+ ...
+ /* jslint ignore:end */
```

Reply:
We should declare the methods directly as `Database.prototype["methodName"]` instead of `Database.prototype.methodName`. And there is no bug here.

Reply:
> for-in loops are slow/expensive operations for historical reasons

This is wrong. Iterating over object keys is indeed slower than iterating over an array, for instance, but doing it with for-in is faster, not slower, than doing it with Object.keys.

But if jslint forbids for-in, you can use object.values: `Object.values(this.statements).forEach(stmt => stmt.free())`

Reply:
implemented Object.values() suggestion in commit 722de939cf89fd540b9afc1e8aafc8b6139757c9